core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
-                            @Override
-                            protected boolean validate(ClusterState newState) {
-                                ClusterBlockException blockException = checkBlock(request, newState);
-                                return (blockException == null || !blockException.retryable());
-                            }
-                        }
-                );
+        AsyncSingleAction(Request request, ActionListener<Response> listener) {
+            this.request = request;
+            // TODO do we really need to wrap it in a listener? the handlers should be cheap
+            if ((listener instanceof ThreadedActionListener) == false) {
+                listener = new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.LISTENER, listener);
+            }
+            this.listener = listener;
+        }
 
-            } else {
-                threadPool.executor(executor).execute(new ActionRunnable(listener) {
-                    @Override
-                    protected void doRun() throws Exception {
-                        masterOperation(request, clusterService.state(), listener);
+        public void start() {
+            this.observer = new ClusterStateObserver(clusterService, request.masterNodeTimeout(), logger);
+            doStart();
+        }
+
+        protected void doStart() {
+            final ClusterState clusterState = observer.observedState();
+            final DiscoveryNodes nodes = clusterState.nodes();
+            if (nodes.localNodeMaster() || localExecute(request)) {
+                // check for block, if blocked, retry, else, execute locally
+                final ClusterBlockException blockException = checkBlock(request, clusterState);
+                if (blockException != null) {
+                    if (!blockException.retryable()) {
+                        listener.onFailure(blockException);
+                    } else {
+                        logger.trace("can't execute due to a cluster block, retrying", blockException);
+                        retry(blockException, retryableOrNoBlockPredicate());
                     }
-                });
-            }
-        } else {
-            if (nodes.masterNode() == null) {
-                if (retrying) {
-                    listener.onFailure(new MasterNotDiscoveredException());
                 } else {
+                    ActionListener<Response> delegate = new ActionListener<Response>() {
+                        @Override
+                        public void onResponse(Response response) {
+                            listener.onResponse(response);
+                        }
+
+                        @Override
+                        public void onFailure(Throwable t) {
+                            if (t instanceof Discovery.FailedToCommitClusterStateException
+                                    || (t instanceof EsRejectedExecutionException && t.getMessage().startsWith("no longer master."))) {
```

Comment:
lets' introduce a dedicated exception for this. We can upgrade discovery.zen.NotMasterException to be in the cluster package and use that.